vars/stageScanForSnyk.groovy

```diff
@@ -0,0 +1,46 @@
+def call(def context, def snykAuthenticationCode, def buildFile, def organisation) {
+  if (!snykAuthenticationCode) {
+    println("Skipping Snyk Scan due to missing authentication code (to enable pass as parameter a snyk service account authentication code to this stage)")
+    return
+  }
+  if (!organisation) {
+    organisation = context.projectId
+    println("organisation set to ${organisation}")
+  }
+
+  String message = "Snyk scan mode: build will " + (context.failOnSnykScanVulnerabilities ? "" : "not ") +
+  "fail if vulnerabilities are found (failOnSnykScanVulnerabilities=${context.failOnSnykScanVulnerabilities})!"
+  println(message)
+
+  if (buildFile == null) {
+    error "build file definition for snyk security scan is null!"
+  } else {
+    stage('Snyk Security Scan') {
+      withEnv(["SNYK_AUTHENTICATION_CODE=${snykAuthenticationCode}", "PROJECT_NAME=${context.targetProject}",
+               "COMPONENT_NAME=${context.componentId}", "BUILD_FILE=${buildFile}", "ORGANISATION=${organisation}"]) {
+        // Verify that snyk is installed
+        def status = sh(script: "snyk version", returnStatus: true)
+        if (status != 0) {
+          error "snyk is not installed!"
+        }
+        // Authorise snyk
+        status = sh(script: "snyk auth $SNYK_AUTHENTICATION_CODE", returnStatus: true)
+        if (status != 0) {
+          error "something went wrong by authorising snyk (SNYK_AUTHENTICATION_CODE=$SNYK_AUTHENTICATION_CODE)!"
```

Comment:
Would it be possible to capture stderr or so? Otherwise the error message is quite cryptic ...

Reply:
I did many test based on the sh command docs: https://jenkins.io/doc/pipeline/steps/workflow-durable-task-step/#code-sh-code-shell-script but I couldn't capture stderr or stdout.

Reply:
yes, I think I found a solution

Reply:
From https://jenkins.io/doc/pipeline/steps/workflow-durable-task-step/#sh-shell-script is not possible to capture stderr, however "Standard error, if any, will still be printed to the log."

sh can return either status (param returnStatus) or stdout (returnStdout). One could opt to capture stdout but then it will be required to parse stdout to decide if command was successfull or not. As mentioned above.

I think because Standard error is printed to the log if any (and after testing and verifying this) that in case of failed build developers will be able to figure out what happened just by looking at the build logs.